### ISSUE TITLE
Leaderboard not well ordered

### ISSUE THREAD

**Issue by Didayolo** [codalab/codabench#1171]
The leaderboard the validation phase of the following competition is not well ordered:

https://www.codabench.org/competitions/1219/

![20230922](https://github.com/codalab/codabench/assets/11784999/44b155ca-68cc-44f2-aa2c-e53aac055f99)

(the competition is private, if you need the secret URL please ask me)

---

UPDATE:

This problem is linked to another problem: the score in the tab "Submissions" (either participant tab or admin tab) does not reflect perfectly the score in the leaderboard.

This is likely due to the javascript rounding found in [submission_manager.tag](https://github.com/codalab/codabench/blob/develop/src/static/riot/competitions/detail/submission_manager.tag):

```javascript
        self.get_score = function (submission) {
            try{
                return parseFloat(submission.scores[0].score).toFixed(2)
           [...]
```

It seems that javascript standard for floats is not precise. The order is based on the scores that you can find in the "Submissions" tab, not the actual scores. The leaderboard reflect the true score.

---

So, it is ordered by primary column, but using a bad rounding.

**Comment by Didayolo** [comment]
Even clicking on "Score" to re-order the rows does not give the right order

**Comment by ihsaan-ullah** [comment]
I think this is not competition specific problem.

In this competition: https://www.codabench.org/competitions/984/#/results-tab
you can see there is a problem 

DESC
<img width="215" alt="Screenshot 2023-09-22 at 5 25 34 PM" src="https://github.com/codalab/codabench/assets/13259262/d1352343-e38b-4a1b-85b4-e8043fcc3bd6">

ASC
<img width="213" alt="Screenshot 2023-09-22 at 5 25 25 PM" src="https://github.com/codalab/codabench/assets/13259262/756f9072-74ea-4bb9-8442-f333f7ba1614">

**Comment by Didayolo** [comment]
I don't understand where the order comes from. It could be based on another score, or on the best score of the participant, maybe on the date... Maybe we should investigate the code directly.

**Comment by ihsaan-ullah** [comment]
> I don't understand where the order comes from. It could be based on another score, or on the best score of the participant, maybe on the date... Maybe we should investigate the code directly.

When you load the leaderboard, the order seems correct. I think the order is changed and is buggy because of the reordering when you click the column 
<img width="208" alt="Screenshot 2023-09-22 at 9 01 57 PM" src="https://github.com/codalab/codabench/assets/13259262/4c111ab9-50cd-471e-9bec-eca18dde5b6e">

**Comment by Didayolo** [comment]
No, in my case, the leaderboard is wrong when you load the page. You can see that the medals (1st, 2nd, etc.) are not well ordered as well.

**Comment by Didayolo** [comment]
UPDATE:

This problem is linked to another problem: the score in the tab "Submissions" (either participant tab or admin tab) does not reflect perfectly the score in the leaderboard.

This is likely due to the javascript rounding found in [submission_manager.tag](https://github.com/codalab/codabench/blob/develop/src/static/riot/competitions/detail/submission_manager.tag):

```javascript
        self.get_score = function (submission) {
            try{
                return parseFloat(submission.scores[0].score).toFixed(2)
           [...]
```

It seems that javascript standard for floats is not precise. The order is based on the scores that you can find in the "Submissions" tab, not the actual scores. The leaderboard reflect the true score.

**Comment by Didayolo** [comment]
In https://github.com/codalab/codabench/blob/develop/src/apps/api/serializers/leaderboards.py :

```python
    def get_submissions(self, instance):
        # desc == -colname
        # asc == colname
        primary_col = instance.leaderboard.columns.get(index=instance.leaderboard.primary_index)
        ordering = [f'{"-" if primary_col.sorting == "desc" else ""}primary_col']
        submissions = Submission.objects.filter(
            phase=instance,
            has_children=False,
            is_specific_task_re_run=False,
            leaderboard__isnull=False, ) \
            .select_related('owner').prefetch_related('scores') \
            .annotate(primary_col=Sum('scores__score', filter=Q(scores__column=primary_col)))

        for column in instance.leaderboard.columns.exclude(id=primary_col.id, hidden=False).order_by('index'):
            col_name = f'col{column.index}'
            ordering.append(f'{"-" if column.sorting == "desc" else ""}{col_name}')
            kwargs = {
                col_name: Sum('scores__score', filter=Q(scores__column__index=column.index))
            }
            submissions = submissions.annotate(**kwargs)

        submissions = submissions.order_by(*ordering, 'created_when')
        return SubmissionLeaderBoardSerializer(submissions, many=True).data
```

**Comment by bbearce** [comment]
Not seeing this right now and I'm not sure what the difference is:

fresh page load:
![image](https://github.com/codalab/codabench/assets/7649007/13233201-a2a7-461c-9841-29c52085aba0)

click Accuracy:
![image](https://github.com/codalab/codabench/assets/7649007/2adfd846-846b-442d-94e4-8f97a7df95ca)

click Accuracy again:
![image](https://github.com/codalab/codabench/assets/7649007/22b2b1da-2b94-4e83-8733-1b3ce2e42bec)

local:
commit 1c3d36618

bbearce@pop-os:~/Documents/codabench$ git branch
* develop

**Comment by bbearce** [comment]
That being said I wonder if we should make a leaderboard issue to encapsulate them all:
* https://github.com/codalab/codabench/issues/1171
* https://github.com/codalab/codabench/issues/1055
* https://github.com/codalab/codabench/issues/873
* https://github.com/codalab/codabench/issues/406

I have one that's new and I can make it its own issue:
While order above isn't my issue I see that computations aren't working right and options I specifically unset and saved persist when I go back into the editor:

Before:
![image](https://github.com/codalab/codabench/assets/7649007/a0d10765-ddd6-4dcc-bc9b-d8382a9a1ad6)

After:
![image](https://github.com/codalab/codabench/assets/7649007/4d89e4d6-c69d-4b3f-aca3-655bdcd83a14)

**Comment by bbearce** [comment]
Hi all,

Picking up where we left off the new test was to make a basic scoring program that copies scores from submission directly to results.txt in the right leaderboard format. This is to make is to I can submit direct scores and have them show up on the leaderboard exactly how I want them. With that said I can recreate the leaderboard from the first image posted about this issue.
![image](https://github.com/codalab/codabench/assets/7649007/843dd908-353d-4e99-8d1c-3cde8af5cffc)
These were the input text files:
![image](https://github.com/codalab/codabench/assets/7649007/a4fa31b9-d6d5-4732-88fd-c1210433fd84)

**Comment by bbearce** [comment]
It's on my localhost. I can share but really any competition will do. It's mainly the scoring program and leaderboard. We can make it live somewhere but I can zip up some data and share here perhaps.
[template_v2_bundle.zip](https://github.com/codalab/codabench/files/12919363/template_v2_bundle.zip)

**Comment by Didayolo** [comment]
OK. It is interesting to note that this does not reproduce the bug. Maybe we need a V1 bundle? Or maybe the problem is really specific to my competition and hard to reproduce.

**Comment by bbearce** [comment]
I can try with a v1 bundle if you want. I was going to focus on the "clicking" of a column that sorts so that perhaps we could make sure  sorting works.

**Comment by Didayolo** [comment]
You are right, at least this is reproducible, and maybe the cause of other problems. It would be good to fix this problem anyway.

**Comment by bbearce** [comment]
PR [1193](https://github.com/codalab/codabench/pull/1193) hopefully resolves this.

**Comment by Didayolo** [comment]
I close this issue for now:

- #1193 resolves the "clicking" order problem
- The general order problem was never reproduced, even with the same competition with a lot of submission

If it happens again, we can re-open the issue or open a new one.

**Comment by Didayolo** [comment]
Let's keep only this issue open to track the problem. 

Also reported here:
- #1389
- #1245

And on the ISPRS International Contest of Individual Tree Crown (ITC) Segmentation: https://www.codabench.org/competitions/3040

**Comment by Didayolo** [comment]
I close this issue to merge and keep track of it here: #1540.

**Comment by Didayolo** [comment]
Happened again

> Yesterday a participant realized the Leaderboard was wrong, as it is suddenly not ranking the participants correctly. I haven’t changed the ranking scheme or the code:

![Image](https://github.com/user-attachments/assets/9af3de67-d6d4-4b28-8b5b-90e9bfec80e6)

We still don't know how to reproduce this problem
